### ISSUE TITLE
Inconsistent ordering of policy rules when adding an Ingress path

### ISSUE THREAD

**Issue by conkiztador** [F5Networks/k8s-bigip-ctlr#1737]
#### Setup Details
 CIS Version : 2.X      
 Build: f5networks/k8s-bigip-ctlr:latest       
 BIGIP Version: Big IP x.x.x      
 AS3 Version: 3.x      
 Agent Mode: AS3       
 Orchestration: K8S      
 Orchestration Version:       
 Pool Mode: Nodeport      

#### Description  

According to the Ingress docs (https://kubernetes.io/docs/concepts/services-networking/ingress/#multiple-matches):

> In some cases, multiple paths within an Ingress will match a request. In those cases precedence will be given first to the longest matching path. If two paths are still equally matched, precedence will be given to paths with an exact path type over prefix path type.

The F5 controller uses the "first matching" strategy for the vserver policy, so the ordering of rules matter. When an Ingress is first created the rules seem to be ordered appropriately, but if more services are added to the Ingress later, they are just added to the end of the policy.

This is a problem particularly for things like `cert-manager`, which adds a temporary path like `/.well-known/acme-challenge/xxxx` to an Ingress for http01 challenge verification.

#### Steps To Reproduce

Example setup:
```apiVersion: v1
kind: Pod
metadata:
  labels:
    app: nginx1
  name: nginx1
spec:
  containers:
  - image: nginx
    name: nginx
    args:
      - "echo one > /usr/share/nginx/html/index.html; exec nginx -g 'daemon off;'"
    command:
      - bash
      - -c
    ports:
    - containerPort: 80
      protocol: TCP
---
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: nginx2
  name: nginx2
spec:
  containers:
  - image: nginx
    name: nginx
    args:
      - "mkdir /usr/share/nginx/html/two; echo two > /usr/share/nginx/html/two/index.html; exec nginx -g 'daemon off;'"
    command:
      - bash
      - -c
    ports:
    - containerPort: 80
      protocol: TCP
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: nginx1
  name: nginx1
spec:
  ports:
  - name: http
    nodePort: 31234
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx1
  type: NodePort
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: nginx2
  name: nginx2
spec:
  ports:
  - name: http
    nodePort: 31235
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx2
  type: NodePort
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: nginx
  annotations:
    ingress.kubernetes.io/ssl-redirect: "false"
    ingress.kubernetes.io/allow-http: "true"
spec:
  rules:
  - host: test.example.com
    http:
      paths:
      - path: /
        backend:
          serviceName: nginx1
          servicePort: 80
      #- path: /two
      #  backend:
      #    serviceName: nginx2
      #     servicePort: 80
```
Apply the above manifest, then later enable the second Ingress path and apply again.

#### Expected Result

Requests to `http://test.example.com/two` should go to `nginx2`.

#### Actual Result  

`nginx2` doesn't receive any traffic.

**Comment by trinaths** [comment]
@conkiztador - End point policy strategy changes as per the need. So suggest use override as3 as below. I verified the same with the above data. it works fine.

```
kind: ConfigMap
apiVersion: v1
metadata:
  name: override-to-best-match
  namespace: default
  labels:
    f5type: virtual-server
    overrideAS3: "true"
data:
  template: |
      {
         "declaration": {
               "test": {
                  "Shared": {
                     "Ingress_172_16_3_132_80": {        <== virtual name
                         "class": "Endpoint_Policy",
                         "strategy": "best-match"         <== strategy.
                     }
                  }
               }
         }
      }
```

configure CIS deployment as below.

```
    spec:
      containers:
      - args:
        .  .  .  .  .
        .  .  .  .  .
        - --override-as3-declaration=default/override-to-best-match   # <namespace>/name-of-override-as3-configMap>
        .  .  .  .  .
        .  .  .  .  .
```

Refer docs for more info on [Override AS3 ConfigMaps](https://clouddocs.f5.com/containers/latest/userguide/config-map.html#override-as3-configmap). 

closing this issue.

**Comment by conkiztador** [comment]
Thanks for the workaround, but shouldn't the f5 agent just correctly follow the K8s Ingress spec without needing to manually set the strategy for every vserver?

I don't think it's reasonable to expect every user of this project to apply a workaround just to get Ingresses to work correctly. Please consider re-opening this issue.

**Comment by trinaths** [comment]
@conkiztador Sure. CIS applies first-match and based on RootApp reorders the rules for best-match. We will reverify this functionality. Creating JIRA CONTCNTR-2534 for PM tracking. Please use override AS3 in the meantime.

**Comment by trinaths** [comment]
Fixed in 2.4.1. Closing this issue.
